docs/defaulting.md

```diff
@@ -0,0 +1,249 @@
+      <mac address='52:54:00:2e:84:30'/>
+      <source network='default'/>
+      <model type='rtl8139'/>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
+    </interface>
+    <input type='mouse' bus='ps2'/>
+    <input type='keyboard' bus='ps2'/>
+    <memballoon model='virtio'>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
+    </memballoon>
+  </devices>
+</domain>
+```
+
+Finally, when a Domain was successfully started by libvirt, libvirt assigns an
+`alias` to every device in the XML.
+
+## The Problem
+
+ * Hotplug operations are difficult with the minimal VM spec of kubevirt,
+   since we could match wrong devices.
+ * After VM restarts, different devices or addresses could be assigned to a VM
+   by libvirt
+ * libvirt assigns all that data on the host, either when defining the Domain,
+   or after the Domain was started. That requires asynchronous lookups and in
+   the worst case even merge logic.
+ * If we don't want to change Guest OS relevant fields in the VM specification
+   behind the back of the user, all computed default values by libvirt, need to
+   be reflected in the VM status.
+
+If your VMs are ephemeral (no migrations, recreating VMs vs. restarting VMs, no
+hotplug), you are not affected by that.
+
+## Proposed Solution
+
+### Defaulting service based on libvirt
+
+Except for some runtime information, which is normally not included in
+migratable domain XMLs, libvirt fills in all defaults when defining a domain,
+which influence how the emulated machine looks like for the Guest OS. We can
+leverage that with a simple REST based service, which defines VMs in an
+isolated libvirt (running libvirt and the small REST service in a Pod). The
+service is a standalone component in KubeVirt and can be integrated in the
+following flow:
+
+ * When doing a POST to
+   `/apis/kubevirt.io/v1alpha1/namespaces/mynamespace/vms`, `virt-api` can do
+   a roundtrip to that service and prefill the defaults. As the response of
+   the POST to `virt-api`, you will get the completely prefilled VM
+   specification. This updated spec contains all the necessary bits, to allow
+   consistent hotplugging and keeping the VM definition consistent between
+   redefines.
+
+Additional flows, like a dedicated `dry-run` endpoint, to allow management
+applications, reusing existing admission controller logic for prefilling VMs,
+withouth having to actually run a VM are thinkable. That would allow complete
+decoupling of managment and runtime.
+
+Defaulting rules:
+
+ * The Kubevirt VM spec can contain abstract definitions, like a reference to a
+   volume claim, or a reference to a host network. In such cases most of the
+   time, the device `source` section needs to be replaced by a dummy `source`
+   inside the defaulter. Then the fully populated `target` section needs to be
+   mapped back (e.g. a PVC does not have to exist, when posting a VM. It can be
```

Comment:
Libvirt would expand the disk to something like this:

```xml
    <disk type='network' device='disk' snapshot='external'>
      <driver name='qemu' type='raw' cache='none'/>
      <source protocol='iscsi' name='iqn.2017-01.io.kubevirt:sn.42/2'>
        <host name='10.107.27.188' port='3260'/>
      </source>
      <backingStore/>
      <target dev='sda' bus='virtio'/>
      <alias name='virtio-disk0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
    </disk>
```

However, the only interesting parts for us are:

```
      <target dev='sda' bus='scis'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
```

so the result would look like this:

```yaml
spec:
  domain:
    devices:
      disks:
      - type: PersistentVolumeClaim
        - source:
          name: disk-01
        - target:
          bus: scis
          device: sda
          address:
             type: pci
             domain: 0x00
             bus: 0x00
             slot: 0x04
             function: 0x00
```

We might also want to retrieve the cache, which is `none` by default, however, that also depends on if we treat the absence of `cache` as `none`, or if we want to default it to `none`. But it could also be that we have cases where we want the cache to be influenced by a PVC. In that case we would never default it anywhere. Anyway, the cache value is not important for the Guest OS and therefore does not need to be defaulted by this libvirt service.